File: 907. Sum of Subarray Minimums/dp.js
/**
 * @param {number[]} arr
 * @return {number}
 */
 var sumSubarrayMins = function(arr) {
    
    const lastSmaller = [];
    const stack = [];
    const n = arr.length;
    const mod = Math.pow(10, 9) + 7
    
    const top = () => stack[stack.length - 1];
    
    // Usinig monotonic stack, find the last smaller number
    for(let i = 0; i < n; i++){
        while(stack.length && arr[top()] >= arr[i]) stack.pop();
        
        if(stack.length){
            lastSmaller[i] = top();
        }else{
            lastSmaller[i] = -1;
        }
        
        stack.push(i)
    }
    
    const dp = [...Array(n)].fill(0);
    
    for(let i = 0; i < n; i++){
        if(lastSmaller[i] === -1){
            const len = i - 0 + 1;
            dp[i] = (arr[i] * len) % mod;
        // There is some smaller number in front of this current number
        }else{
            // Find the length from i to lastSmaller + 1;
            const lenWithI = i - lastSmaller[i];
            dp[i] = (dp[lastSmaller[i]] + (lenWithI * arr[i])) % mod;
        }
    }
    
    return dp.reduce((acc, num) => acc + num, 0) % mod;
};

File: 0907. Sum of Subarray Minimums/dp.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        
        int N = arr.length;
        int[] dp = new int[N];
        Arrays.fill(dp, 0);
        Deque<Integer> stack = new ArrayDeque<>();
        int total = 0;
        int mod = (int)(Math.pow(10, 9) + 7);
        
        for(int i = 0; i < N; i++){
            // Find the position of the last smaller number
            while(!stack.isEmpty() && arr[stack.peek()] >= arr[i]) stack.pop();
            
            if(stack.size() == 0){
                int len = i + 1;
                dp[i] = (arr[i] * len) % mod;
            }else{
                
                int lastSmaller = stack.peek();
                dp[i] = (dp[lastSmaller] + (arr[i] * (i - lastSmaller))) % mod;
            }
            
            
            total += dp[i] % mod;
            stack.push(i);
        }
        
        return total % mod;
    }
}